### ISSUE TITLE
container creation fails intermittently with "unable to set hairpin mode"

### ISSUE THREAD

**Issue by bprodoehl** [docker/docker#15855]
I am running the daemon with --userland-proxy=false to get past some memory leaks and other problems associated with having some containers expose relatively large numbers of ports, and now I'm seeing intermittent container start calls fail with "unable to set hairpin mode".  Plenty of these create-start operations succeed, with the same container under essentially the same conditions.

```
# uname -a
Linux myhost 3.13.0-62-generic #102-Ubuntu SMP Tue Aug 11 14:29:36 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```
```
# docker info
Containers: 145
Images: 118
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 408
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-62-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 2
Total Memory: 1.947 GiB
Name: myhost
ID: DEAD:BEEF
Username: myuser
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```
```
# docker version
Client:
 Version:      1.9.0-dev
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   88cacbf
 Built:        Mon Aug 24 15:36:20 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0-dev
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   88cacbf
 Built:        Mon Aug 24 15:36:20 UTC 2015
 OS/Arch:      linux/amd64
```

```
INFO[37408] POST /containers/create?HostConfig=&Env=MYENV%3Dvalue&Image=my%2Fimage%3Atag&name=mycontainername
INFO[37408] GET /_ping                                   
INFO[37408] POST /containers/31ebde3a2b09135c01d2614cb5126f6cca4a344ed0451832c768a07bb2281a2d/start 
ERRO[37408] Handler for POST /containers/{name:.*}/start returned error: Cannot start container 31ebde3a2b09135c01d2614cb5126f6cca4a344ed0451832c768a07bb2281a2d: unable to set hairpin mode on veth001565b via netlink: operation not supported 
ERRO[37408] HTTP Error                                    err=Cannot start container 31ebde3a2b09135c01d2614cb5126f6cca4a344ed0451832c768a07bb2281a2d: unable to set hairpin mode on veth001565b via netlink: operation not supported statusCode=500
```

**Comment by bprodoehl** [comment]
I'm thinking this will be more common once #14973 lands.

**Comment by bprodoehl** [comment]
Oddly enough, I don't see anything in syslog about the failing veths.  I see this on a successful container creation, but nothing when I grep syslog or dmesg for one that fails.  So, this is from a successful start, but it still looks fishy.

```
Aug 26 16:20:17 host kernel: [15168005.492582] device veth02f4f78 entered promiscuous mode
Aug 26 16:20:17 host kernel: [15168005.493292] IPv6: ADDRCONF(NETDEV_UP): veth02f4f78: link is not ready
Aug 26 16:20:17 host kernel: [15168005.640832] IPv6: ADDRCONF(NETDEV_CHANGE): veth02f4f78: link becomes ready
Aug 26 16:20:17 host kernel: [15168005.640872] docker0: port 10(veth02f4f78) entered forwarding state
Aug 26 16:20:17 host kernel: [15168005.640877] docker0: port 10(veth02f4f78) entered forwarding state
Aug 26 16:20:18 host kernel: [15168006.241029] IPv6: eth0: IPv6 duplicate address fe80::42:acff:fe11:16 detected!
Aug 26 16:20:20 host ntpd[2310]: Listen normally on 32 veth02f4f78 fe80::8c89:88ff:fe27:5eb2 UDP 123
Aug 26 16:20:20 host ntpd[2310]: peers refreshed
Aug 26 16:20:20 host ntpd[2310]: new interface(s) found: waking up resolver
Aug 26 16:20:32 host kernel: [15168020.661368] docker0: port 10(veth02f4f78) entered forwarding state
Aug 26 16:20:38 host kernel: [15168026.304449] docker0: port 10(veth02f4f78) entered disabled state
Aug 26 16:20:38 host kernel: [15168026.444581] docker0: port 10(veth02f4f78) entered disabled state
Aug 26 16:20:38 host kernel: [15168026.445464] device veth02f4f78 left promiscuous mode
Aug 26 16:20:38 host kernel: [15168026.445476] docker0: port 10(veth02f4f78) entered disabled state
Aug 26 16:20:40 host ntpd[2310]: Deleting interface #32 veth02f4f78, fe80::8c89:88ff:fe27:5eb2#123, interface stats: received=0, sent=0, dropped=0, active_time=20 secs
Aug 26 16:20:40 host ntpd[2310]: peers refreshed
```

I'm wondering if this might be a different face of #15341.

**Comment by bprodoehl** [comment]
Can also reproduce with a pretty simple case.

```
# docker run --rm ubuntu echo hello!
hello!
# docker run --rm ubuntu echo hello!
hello!
# docker run --rm ubuntu echo hello!
Error response from daemon: Cannot start container 943fe6c0f381503647c00fa1a0233f57c280840b1c4fa906f751fedec2583e48: unable to set hairpin mode on vethac139fe via netlink: operation not supported
# docker run --rm ubuntu echo hello!
hello!
```

One of my test hosts is reproducing this problem very reliably, and I have yet to reproduce the problem on my other test host.  I'm going to add a couple more hosts to the mix.

**Comment by mavenugo** [comment]
@bprodoehl yes. it seems like a different face of #15431. can you join us in #docker-network channel and we can debug it together in real time.

**Comment by bprodoehl** [comment]
Thanks for the offer. I'm there now.

**Comment by crosbymichael** [comment]
this has been fixed my #16038 thanks!

**Comment by vcaputo** [comment]
FYI I'm seeing this issue on docker 1.12.1 w/go 1.6.3 on CoreOS (dev build, alpha 1153.0.0 is similar) running the following test:
```
i=0; while echo $i && docker run --rm -p 8080 busybox /bin/true && docker ps; do sleep 0.05; ((i+=1)); done
```
with  `--userland-proxy=false`

I'm attempting to repro `https://github.com/coreos/bugs/issues/254` but this issue is triggering far more readily and getting in the way.

**Comment by thaJeztah** [comment]
@vcaputo could you open a new issue? This issue was on an older version of docker, and possibly there's a new issue (with the same effect)
